### DISCUSSION TITLE
How to embed script into Interpreter

### DISCUSSION THREAD

**Discussion by SebastianMpl** [Oldes/Rebol3#69] [Q&A]
Hi there,
Thank you very much for your Rebol fork. It's great and saves a lot of work!

The thing that I want to ask is that I have script that I can run with Rebol interpreter, but is it possible to somehow embed the script into the rebol executable file so that it could be run as an executable? 
I found out that for Carl's Rebol there was something called encap:
http://www.rebol.com/docs/sdk/encap.html

Is something similar available here?

What to look for?

Thank you,

**Comment by Oldes** [comment]
Hi... it depends how much skills do you have?
If I remember it well (I was not using the old Rebol SDK), it just merged multiple scripts together and then somehow attached the single script to the app binary, so it was done without need to compile anything. That is not supported for now.

But if you are able to compile own Rebol, then it is rather easy. Good (and only) example how to do it is the [Siskin Builder](https://github.com/Siskin-framework/Builder) utility, which I use to build Rebol itself. It is actually a customised Rebol.

It is compiled using this specification file: https://github.com/Siskin-framework/Builder/blob/master/tree/rebol/siskin.nest

If you read the file, you can see, that there is a section with [some parts of the Rebol sources](https://github.com/Siskin-framework/Builder/blob/17fd6e6b16ffcd741679f1f5ecb732d74ef7d50f/tree/rebol/siskin.nest#L18-L24), custom [mezzanine files](https://github.com/Siskin-framework/Builder/blob/17fd6e6b16ffcd741679f1f5ecb732d74ef7d50f/tree/rebol/siskin.nest#L26-L32) and an [entry script](https://github.com/Siskin-framework/Builder/blob/17fd6e6b16ffcd741679f1f5ecb732d74ef7d50f/tree/rebol/siskin.nest#L34).

Also quite useful are [these 2 defines](https://github.com/Siskin-framework/Builder/blob/17fd6e6b16ffcd741679f1f5ecb732d74ef7d50f/tree/rebol/siskin.nest#L40-L41), to let the compiler know, that we are not building just an ordinary Rebol, but a custom app.

I don't think it is perfect... but it works for me so far and I can use [GitHub action](https://github.com/Siskin-framework/Builder/blob/master/.github/workflows/siskin.yml) to build binaries for all main systems directly from web.

Things to be improved in the current process:
1. instead of including multiple scripts like it is now, there must be a way, how to preprocess these into the single file, like it was done with the SDK.
2. it is possible to have Rebol as a library and the utilities could be just like the host files.. I still try to keep this possibility alive.. for example [lib-rebol3-core-windows-x64.dll](https://github.com/Oldes/Rebol3/releases/download/3.10.3/lib-rebol3-core-windows-x64.dll) and [host-core-windows-x64.exe](https://github.com/Oldes/Rebol3/releases/download/3.10.3/host-core-windows-x64.exe), where the exe is only 37kB and the library could be shared between multiple apps.

There was one day also an extension for embedding scripts to R3, but I am not sure, how much open sourced it was and where the sources are.

**Comment by SebastianMpl** [comment]
> Hi... it depends how much skills do you have? If I remember it well (I was not using the old Rebol SDK), it just merged multiple scripts together and than somehow attached the single script to the app binary, so it was done without need to compile anything. That is not supported for now.
> 

Thank you  for quick reply. Indeed I was looking for something like you said in the beginning - that works without compilation. But since it's currently not supported I'll dive deeper into Siskin Builder with the help of links that You provided.

And as for the extention you mentioned in the end I also found that link is dead.

Thank you very much :)
